File: lib/ex_llama/model_options.ex
defmodule ExLLama.ModelOptions do
  defstruct [
    :n_gpu_layers,
    :split_mode,
    :main_gpu,
    :vocab_only,
    :use_mmap,
    :use_mlock
  ]

  @type t :: %__MODULE__{
               n_gpu_layers: non_neg_integer(),
               split_mode: String.t, #  :none | :layer | :row,
               main_gpu: non_neg_integer(),
               vocab_only: boolean(),
               use_mmap: boolean(),
               use_mlock: boolean()
             }

  def new() do
    %__MODULE__{
      n_gpu_layers: 0,
      split_mode: "none",
      main_gpu: 0,
      vocab_only: false,
      use_mmap: false,
      use_mlock: false
    }
  end
end

File: native/erlang_llama_cpp_nif/src/structs/model_options.rs
use llama_cpp::{LlamaParams, SplitMode};
use rustler::{NifStruct};
//use rustler::types::atom;

#[derive(NifStruct)]
#[module = "ExLLama.ModelOptions"]
pub struct ModelOptions {
    pub n_gpu_layers: u32,
    pub split_mode: String,
    pub main_gpu: u32,
    pub vocab_only: bool,
    pub use_mmap: bool,
    pub use_mlock: bool,
}

impl From<ModelOptions> for LlamaParams {
    fn from(value: ModelOptions) -> Self {
        // Map the string values to the corresponding SplitMode enum values
        let split_mode = match value.split_mode.as_str() {
            "none" => SplitMode::None,
            "layer" => SplitMode::Layer,
            "row" => SplitMode::Row,
            _ => panic!("Invalid split_mode value"),
        };

        Self {
            n_gpu_layers: value.n_gpu_layers,
            split_mode: split_mode,
            main_gpu: value.main_gpu,
            vocab_only: value.vocab_only,
            use_mmap: value.use_mmap,
            use_mlock: value.use_mlock,
        }
    }
}